src/entities/modelData.ts

```diff
@@ -57,29 +57,22 @@ export class ModelData {
      * Sets a particular component of this model.
      */
     public setComponent<T>(component: ModelComponent, value: T): void {
-        console.log("set component is called");
-        console.log(this.components.getValue(component));
-        console.log(value);
-        if (value !== this.components.getValue(component)) {
-            console.log("setted component", component);
+        // TODO ask Jonathan why this check was here
+        // This causes that changeBuffer is not always updated correctly
+        // (e.g. you create component, set it, update the component and set it again:
+        // changeBuffer is not updated second time
+
+        // if (value !== this.components.getValue(component)) {
             this.changeBuffer.add(component);
```

Comment:
This check was a feature, not a bug. The bug, conceptually, is that the model implicitly assumes that `setComponent<T>` is the only way to change the model data, i.e., that model components are *immutable.*

A consequence of this assumption is that `setComponent(ModelComponent.DataGraph, getComponent(ModelComponent.DataGraph));` is a no-op. So there's really no need to pretend like a component has changed when a situation like this occurs. In fact, spurious "update" notifications are a misfeature because they trigger unnecessary new work (which may in turn trigger more work&mdash;you get the picture).

What I'm assuming you're trying to achieve by deleting this check is you're trying to notify the model that a *mutable component* has changed when in fact the model assumes that all components are *immutable.* Now the check you deleted is just a consequence of the immutability assumption (which actually holds for, say, validation results). So this edit kind of treats the symptoms instead of the disease.

I initially thought there might be a quick fix for this mismatch between assumption and reality, like deleting this check or (better) introducing a separate function to tell the model that a mutable component has changed. But I'm no longer convinced that is the case because out-of-order execution is also based on the immutability assumption. Deleting this check now will only cause more Heisenbugs later on.

I am truly sorry for failing to consider this unfortunate situation before. I guess there are three ways to fix this:
  1. Switch to all-immutable model components. This is actually the best solution but `rdflib.js`'s graph data structure is sadly mutable so this solution is probably not very realistic.
  2. Modify the model to take into account mutable data structures by inserting barriers. This solution is extremely likely to F up the benefits we get from out-of-order execution in a big way.
  3. Come up with some fancy new concept that'll somehow allow the model to contain the mutability without copying entire objects. This option is a hail mary at its best and a at least as bad as option one at its worst.

I'll try to think about this issue for the next few days. In the meanwhile, restore the check and add a function to explicitly tell the model that something inside of a component has changed. Use that to signal changes instead of using `setComponent` for purposes outside of its domain.

Reply:
I created a method in ModelData to to explicitly tell when a component has changed. But in that method I needed to change the changeBuffer for obvious reasons. To realize that I stopped using the clone method for the ModelData since otherwise I did not have access to the changeBuffer. 

This looked like the best quickfix I could do for now ...